PROTOCOL.md

```diff
@@ -0,0 +1,347 @@
+array of the hash of each required item when encrypted under `kₗ`
+(`eitemhash`).
+6. Finally, the `tenant` returns the encrypted items as well as the keys used
+to encrypt them (`kₗ`). Note that the `keep` **must** validate that the items
+actually match the hash for the item in the `canon`. This is particularly
+important for the `Multi-Party Computation` model.
+
+## Model #2: Remote Pull
+
+Under the `Remote Push` model we had simple operation. However, we made the
+assumption that the `tenant` who wishes to deploy a workload is also the
+`scheduler` of the `enarx run` process. This may not be the case. For example,
+a `tenant` may wish to wrap `enarx run` in a container and execute it using
+`kubernetes` as a `scheduler`. This requires we break our previous `tenant`
+process into three distinct components:
+
+1. the `operator` who wants to deploy a workload
+2. the `scheduler` that will actually trigger the `enarx run` process (i.e.
+`kubernetes`)
+3. the `targetsrv` that will supply the `keep` with its `items`
+
+We can achieve this using the same protocol as before, but by using a separate
+`targetsrv` process and some additional steps during the `Instantiation Phase`.
+
+### Phase #1: Instantiation
+
+Instantiating a `keep` under the `Remote Pull` model is basically the same as
+the `Remote Push` model, but with some additional steps at the beginning.
+
+```mermaid
+graph LR
+    subgraph tenant
+        A(operator)
+        B(targetsrv)
+    end
+
+    subgraph host
+        C(scheduler)
+        D(enarx run)
+        E(enarx srv)
+        F(keep)
+
+	    style F fill:#0f0,stroke:#333,stroke-width:4px
+    end
+
+    A -. "1. enarx push" .-> B
+    B -. "2. pubₜ" .-> A
+
+    A -. "3. configure [target = (uri, pubₜ)]" .-> A
+    A -. "4. schedule" .-> C
+
+    C -.-> D
+    D --> E
+    E -.-> F
+```
+
+1. The `operator` issues the `enarx push` command to store a fully deployable
+keep inside the `targetsrv`. The `targetsrv` receives the the following data:
+measurement of all allowed `shim`, `exec` and `backend` combinations, the
+`manifest` and all `items`.
+2. Upon receipt of the aforementioned items, the `targetsrv` creates a new keep
+instance record storing all the items and generates the target keypair (`pubₜ`,
+`prvₜ`), returning the public key (`pubₜ`) to the `operator`. The `enarx push`
+process returns an appropriate `target` configuration. This will likely have a
+`uri` of `tcp:...` so that it can connect to the `targetsrv`.
+3. The `operator` uses the `target` configuration to produce `targets` which it
+associates with the job to run. For example, the operator might include the
+`targets` in a `Dockerfile` and then build it with `docker build`.
+4. The `operator` schedules the job to run. For example, this might include
+configuring the aforementioned container to run under `kubernetes`.
+
+All remaining steps are identical to the `Remote Push` model.
```

Comment:
I think this is insufficiently clear.  Need to explain what step in the Remote Push model we now jump to.